### ISSUE TITLE
Inspection Proposal: Warn when depending on final classes

### ISSUE THREAD

**Issue by goetas** [kalessil/phpinspectionsea#287]
Consider this example:

```php
final class Consumer
{
    function consume()
    {
    }
}

class Task
{
    function __construct(Consumer $consumer)
    {
     //...
    }
}
```

Should this be considered as a bad practice? Testing `Task` will be hard because is not possible to mock `Consumer` since is `final`

Can we "suggest" the user to add an interface and depend on it? 
The right code should be:

```php
interface ConsumerInterface
{
    function consume();
}

final class Consumer implements ConsumerInterface
{
    function consume()
    {
    }
}

class Task
{
    function __construct(ConsumerInterface $consumer)
    {
     //...
    }
}
```
There are a lot of good reasons to declare a class as `final` (see https://ocramius.github.io/blog/when-to-declare-classes-final/) but depending on them is an issue IMO.

Feedback?

**Comment by kalessil** [comment]
Thank you for proposal @goetas, can this be bundled together with #265?

**Comment by funivan** [comment]
@kalessil I think you are correct, this feature is part of #265 but with one difference. 
for example:
```php
final class Consumer{
    function consume()   {    }
}

class Task{
    function __construct(Consumer $consumer) //<<Warning should be here
    {
    }
}
```
If we remove `final` keyword from the class than `weak warning` can be produce, or even nothing (user can configure this behavior)

**Comment by goetas** [comment]
Looks similar to #265 but the implementation should be much easier

**Comment by Fleshgrinder** [comment]
> Mocking is a code smell.
>
> — [Greg Young](http://www.codebetter.com/gregyoung/2008/02/13/mocks-are-a-code-smell/)

Making classes `final` is good practice and mocking is definitely bad practice. It often leads to green tests, but the objects are not compatible to each other in reality. Simply using the real object as is will not only be faster, drop the requirement for a mocking framework altogether, no, it will also ensure that the tests are _always_ using the latest version.

**Comment by kalessil** [comment]
@funivan : removing final would make the warning disappear (we can not offer any interface to be placed instead).

**Comment by kalessil** [comment]
@goetas : I just want to bundle related work, if we promoting available interfaces, we can report using final classes. For me, it'll mean easier maintenance.

**Comment by kalessil** [comment]
@Fleshgrinder : agree, it just fits perfectly with promoting the interfaces usage here.

- Final class is a good thing, no doubt
- But consuming code should rather rely on an interface but not a final implementation.

**Comment by kalessil** [comment]
Marking as duplicate of #265
